### PR TITLE
Add tests for expect ident but find enum or struct panic

### LINKED ISSUE #22426: Panic: "internal compiler error: ident only path should have been covered already"
This error is triggered by providing type parameters to a generic type in a `match` pattern.

e.g.
```rust
#![crate_type = "staticlib"]

pub struct Foo<T>(T, T);

impl<T> Foo<T> {
    fn foo(&self) {
        match *self {
            Foo<T>(x, y) => println!("Goodbye, World!")
        }
    }
}
```
[Playpen here.](http://is.gd/qUaQmI)

### LINKED ISSUE #22589: internal compiler error on reading stdin: panicked at Box<Any>
Ok, this is not syntactically correct, but compiler should not panick

Using rustc 1.0.0-nightly (522d09dfe 2015-02-19) (built 2015-02-19)

use std::old_io;
fn main()
{
    let s = old_io::stdin().read_line().ok().expect("");
    if let Result<n2,_> = s.trim().parse() {};
}

### LINKED ISSUE #22647: Compiler panicked "ident only path should have been covered already"
Hello.
Trying to learn Rust, playing with closures. Compiler asked me to report bug.

Code:
```
fn main() {
	let caller<F> = |f: F|
	where F: Fn() -> i32
	{
		let x = f();
		println!("Y {}",x);
		return x;
	};

	caller(bar_handler);
}

fn bar_handler() -> i32 {
	println!("zz");
	return 5;
}
```

Error:

```
 cargo run RUST_BACKTRACE=1
   Compiling app v0.0.1
src/main.rs:2:6: 2:12 error: internal compiler error: ident only path should have been covered already
src/main.rs:2   let caller<F> = |f: F|
                    ^~~~~~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libsyntax/diagnostic.rs:129

```

File `/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libsyntax/diagnostic.rs` doesn't exist, folder `/Users/rustbuild/` too.

In this code I tried to create closure with function as first argument, but can't yet find how to declare it. But it's offtopic :)

### LINKED ISSUE #22665: internal compiler error: ident only path should have been covered already
I tried this code:

    fn main(){
      let line = std::old_io::stdin().read_line().ok().expect("Failed to read line");
      let Result<value : u8,_> = line.parse();
      println!("{}",value)
    }

I expected it to compile or give me a helpful error message.

Instead, this happened: 

    t2.rs:4:7: 4:13 error: internal compiler error: ident only path should have been covered already
    t2.rs:4   let Result<value : u8,_> = line.parse();
                  ^~~~~~
    note: the compiler unexpectedly panicked. this is a bug.
    note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
    note: run with `RUST_BACKTRACE=1` for a backtrace thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:129

## Meta

`rustc --version --verbose`:

    rustc 1.0.0-nightly (522d09dfe 2015-02-19) (built 2015-02-20)
    binary: rustc
    commit-hash: 522d09dfecbeca1595f25ac58c6d0178bbd21d7d
    commit-date: 2015-02-19
    build-date: 2015-02-20
    host: x86_64-unknown-linux-gnu
    release: 1.0.0-nightly

Backtrace:

    stack backtrace:
       1:     0x7fdd80fb5210 - sys::backtrace::write::h252031bd050bf19aKlC
       2:     0x7fdd80fddac0 - panicking::on_panic::h8a07e978260e2c7btXL
       3:     0x7fdd80f1e720 - rt::unwind::begin_unwind_inner::h322bcb3f35268c19RBL
       4:     0x7fdd7e20caa0 - rt::unwind::begin_unwind::h4707836431883763822
       5:     0x7fdd7e20ca30 - diagnostic::SpanHandler::span_bug::hc8a5aaa027e1ffe4vYE
       6:     0x7fdd7e22f440 - parse::parser::Parser<'a>::parse_pat::hb2ecf80c7f350654wLL
       7:     0x7fdd7e2454d0 - parse::parser::Parser<'a>::parse_local::h4dbb2e49605b5f8fA1L
       8:     0x7fdd7e2456f0 - parse::parser::Parser<'a>::parse_stmt::h82edbd6867b9ac1dv4L
       9:     0x7fdd7e24f0a0 - parse::parser::Parser<'a>::parse_block_tail_::h1a3995ee7a4c43efYfM
      10:     0x7fdd7e22d830 - parse::parser::Parser<'a>::parse_inner_attrs_and_block::he49e0d8ec104204cieM
      11:     0x7fdd7e2536e0 - parse::parser::Parser<'a>::parse_item_fn::hbc31e03198152f3cfVM
      12:     0x7fdd7e248370 - parse::parser::Parser<'a>::parse_item_::h60b4bcefd23d6140zUN
      13:     0x7fdd7e2598e0 - parse::parser::Parser<'a>::parse_mod_items::h79ee3c5e427ceaaaPjN
      14:     0x7fdd7e264e30 - parse::parser::Parser<'a>::parse_crate_mod::h221baa0e96bbcf97FlO
      15:     0x7fdd7e277cf0 - parse::parse_crate_from_file::h14706417f2de711evNV
      16:     0x7fdd8163f340 - driver::phase_1_parse_input::closure.16948
      17:     0x7fdd81619670 - driver::phase_1_parse_input::h6ce7c234588dc89eQra
      18:     0x7fdd81616070 - driver::compile_input::h4d3049bece5f3590Gba
      19:     0x7fdd816e52b0 - run_compiler::h691fc217cf3d8d57Zbc
      20:     0x7fdd816e3bb0 - thunk::F.Invoke<A, R>::invoke::h14302181389313704581
      21:     0x7fdd816e2aa0 - rt::unwind::try::try_fn::h15519976819974673451
      22:     0x7fdd8104a880 - rust_try_inner
      23:     0x7fdd8104a870 - rust_try
      24:     0x7fdd816e2da0 - thunk::F.Invoke<A, R>::invoke::h8606420552169060279
      25:     0x7fdd80fc9b60 - sys::thread::thread_start::h3defdaea150d8cd693G
      26:     0x7fdd7aecb2b0 - start_thread
      27:     0x7fdd80ba7249 - __clone
      28:                0x0 - <unknown>

### LINKED ISSUE #22712: Unexpected panic when compiling ill-formed code
When compiling ill formed code the rust compiler panics, printing: "error: internal compiler error: ident only path should have been covered already".
I saw a few issues with the same error message ( #22665 , #22426 , #22589 ) but to me they seem to  be different bugs...

I tried this code:

    use std::vec::Vec;

    struct Foo<B> {
        buffer: B
    }

    fn bar() {
      let Foo<Vec<u8>>
    }

I expected to see this happen: A nice compiler error message.

Instead, this happened: Internal compiler error.

## Meta

`rustc --version --verbose`:

    rustc 1.0.0-nightly (2b01a37ec 2015-02-21) (built 2015-02-22)
    binary: rustc
    commit-hash: 2b01a37ec38db9301239f0c0abcf3c695055b0ff
    commit-date: 2015-02-21
    build-date: 2015-02-22
    host: x86_64-unknown-linux-gnu
    release: 1.0.0-nightly

Backtrace:

    $ RUST_BACKTRACE=1 rustc lib.rs 

    lib.rs:8:7: 8:10 error: internal compiler error: ident only path should have been covered already
    lib.rs:8   let Foo<Vec<u8>>
                   ^~~
    note: the compiler unexpectedly panicked. this is a bug.
    note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
    note: run with `RUST_BACKTRACE=1` for a backtrace
    thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:129

    stack backtrace:
       1:     0x7f8ea8678210 - sys::backtrace::write::h58f3688f47ea9825KlC
       2:     0x7f8ea86a0ac0 - panicking::on_panic::h6a53d3220988f2eebYL
       3:     0x7f8ea85e1720 - rt::unwind::begin_unwind_inner::h72a0c6e103bfcdc3zCL
       4:     0x7f8ea58abde0 - rt::unwind::begin_unwind::h2955836616559227712
       5:     0x7f8ea58abd70 - diagnostic::SpanHandler::span_bug::h3f1d138a57b9309dqZE
       6:     0x7f8ea58ce780 - parse::parser::Parser<'a>::parse_pat::h809bd931d91273bfrML
       7:     0x7f8ea58e4840 - parse::parser::Parser<'a>::parse_local::h2e00e7b160166912v2L
       8:     0x7f8ea58e4a60 - parse::parser::Parser<'a>::parse_stmt::h0653bed7d257c65eq5L
       9:     0x7f8ea58ee410 - parse::parser::Parser<'a>::parse_block_tail_::h3c43c34d39ff87a5TgM
      10:     0x7f8ea58ccb70 - parse::parser::Parser<'a>::parse_inner_attrs_and_block::hebb7a527b0f1d8c5dfM
      11:     0x7f8ea58f2a50 - parse::parser::Parser<'a>::parse_item_fn::hc7a9a7383a82e7a4aWM
      12:     0x7f8ea58e76e0 - parse::parser::Parser<'a>::parse_item_::h9be2ca1f12dcd0aeuVN
      13:     0x7f8ea58f8c50 - parse::parser::Parser<'a>::parse_mod_items::hb94387fccea63b29KkN
      14:     0x7f8ea59041a0 - parse::parser::Parser<'a>::parse_crate_mod::h961b7044a0ca58e2AmO
      15:     0x7f8ea5917060 - parse::parse_crate_from_file::h8fb07d2676fac76cqOV
      16:     0x7f8ea8d025a0 - driver::phase_1_parse_input::closure.16948
      17:     0x7f8ea8cdc8d0 - driver::phase_1_parse_input::h49108518c50c4641Qra
      18:     0x7f8ea8cd92d0 - driver::compile_input::hfb1bb38c59866d0fGba
      19:     0x7f8ea8da8510 - run_compiler::hf2a7d9848e51e3b5Zbc
      20:     0x7f8ea8da6e10 - thunk::F.Invoke<A, R>::invoke::h12698859262113329290
      21:     0x7f8ea8da5d00 - rt::unwind::try::try_fn::h10854600533917764358
      22:     0x7f8ea870d880 - rust_try_inner
      23:     0x7f8ea870d870 - rust_try
      24:     0x7f8ea8da6000 - thunk::F.Invoke<A, R>::invoke::h4174228679526982116
      25:     0x7f8ea868cb60 - sys::thread::thread_start::h60ae7cb39d135ef7R4G
      26:     0x7f8ea256a0c0 - start_thread
      27:     0x7f8ea8259fd9 - __clone
      28:                0x0 - <unknown>

### PR DESCRIPTION
**Pull request by edwardw**
Closes #22589
Closes #22647
Closes #22665
Closes #22712

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -3505,52 +3505,44 @@ impl<'a> Parser<'a> {
...
+                            );
+                            pat = PatEnum(enum_path, Some(args));
+                        }
+                    }
+                    _ => {
+                        if !enum_path.global &&
+                            enum_path.segments.len() == 1 &&
+                            enum_path.segments[0].parameters.is_empty()
+                        {
+                            // We expect an identifier, but duo to
```
Comment (nick29581) [inline src/libsyntax/parse/parser.rs]: nit s/duo/due

**Comment by nick29581** [comment]
r+ with the typo fixed.

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (edwardw) [inline src/libsyntax/parse/parser.rs]: Done.

**Comment by Manishearth** [comment]
When running `check-stage1` on this I get the following error:

```
/opt/rust/src/test/parse-fail/issue-22426.rs:18:16: 18:17 error: unexpected token: `<`
/opt/rust/src/test/parse-fail/issue-22426.rs:18             Foo<T>(ref x, ref y) => {}  //~ ERROR expected identifier, found enum or struct
                                                               ^
```


Should I expect this to only work for `check-stage2`?

**Comment by edwardw** [comment]
This is weird. This test in my local `make check-stage1-pfail` runs fine.

**Comment by Manishearth** [comment]
@bors: r-

I figured it out, https://github.com/rust-lang/rust/pull/22544 broke your PR. Please rebase onto master and ask me for an r+ again (I'll try to include it into the rollup if there's time). Thanks!

**Comment by Manishearth** [comment]
dammit bors

**Comment by edwardw** [comment]
I'm looking into it right now, could be some in between commits caused behavior change of the parser. And building LLVM, it will take a while.

**Comment by Manishearth** [comment]
@edwardw No need to look for it, the change that broke this PR was https://github.com/rust-lang/rust/pull/22544 :)

**Comment by edwardw** [comment]
@Manishearth yeah, but I still need to verify it locally first. Besides I have the feeling that my patch will reduce to just bunch of tests :bicyclist:

**Comment by edwardw** [comment]
It is merely tests now.

r? @Manishearth

**Comment by l0kod** [comment]
Not sure it's exactly the same, but crash Rust nightly from play.rust-lang.org as well:

```rust
use std::rc::Rc;

fn main() {
    match Rc::new(1) {
        Rc<_> => {}
    }
}
```

**Comment by edwardw** [comment]
@l0kod, playpen must've been a little bit lagged behind. Just run your code locally and `rustc` correctly bailed itself out:

```rust
use std::rc::Rc;

fn main() {
    match Rc::new(1) {
        Rc<_> => {}  // error: unexpected token: `<`
    }
}
```

**Pull request merged**